src/lightning/fabric/utilities/spike.py

```diff
@@ -0,0 +1,149 @@
+import json
+import operator
+import os
+import warnings
+from typing import Any, Dict, List, Literal, Optional, Protocol, runtime_checkable, Union
+
+import torch
+from lightning_utilities.core.imports import compare_version
+
+_TORCHMETRICS_GREATER_EQUAL_1_0_0 = compare_version("torchmetrics", operator.ge, "1.0.0")
+if _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+    from torchmetrics.aggregation import MeanMetric
+    from torchmetrics.wrappers import Running
+
+
+class SpikeDetection:
+    def __init__(
+        self,
+        mode: Literal["min", "max"] = "min",
+        window: int = 10,
+        warmup: int = 1,
+        atol: Optional[float] = None,
+        rtol: Optional[float] = 2.0,
+        exclude_batches_path: Optional[str] = None,
+    ):
+        if not _TORCHMETRICS_GREATER_EQUAL_1_0_0:
+            raise RuntimeError("SpikeDetection requires torchmetrics>=1.0.0! Please upgrade your version!")
+        super().__init__()
+
+        self.last_val: Union[torch.Tensor, float] = 0.0
+        # spike detection happens individually on each machine
+        self.running_mean = Running(MeanMetric(dist_sync_on_step=False, sync_on_compute=False), window=window)
+        self.mode = mode
+        self.warmup = warmup
+        self.atol = atol
+        self.rtol = rtol
+        self.bad_batches: List[int] = []
+        self.exclude_batches_path = exclude_batches_path
+
+    @torch.no_grad()
+    def on_train_batch_end(self, fabric: "_BooleanReducer", loss: torch.Tensor, batch: Any, batch_idx: int) -> None:
```

Comment:
Why is fabric annotated like this? I see it everywhere.

Reply:
just so that I can also pass the trainer to it. We don't really need everything from fabric there, just reducing a boolean

Reply:
changed in https://github.com/Lightning-AI/lightning/pull/18014